### PR TITLE
Fix assertion failure due to missing clause id

### PR DESCRIPTION
**Pull request by 4tXJ7f**
This commit fixes bug 821. As written in the description of the bug, the issue
is that `id` is not being set on one of the paths in addClause(), specifically
in the case where all but one literal are assigned false and the remaining
literal is assigned true. In that case, we are not actually adding anything and
set the `id` to `ClauseIdUndef`.

### CODE REVIEW

**Comment by 4tXJ7f** [comment]
Note: I do not know this part of the code base at all and the test case attached to the bug report still fails but with a different assertion failure and much later. Please review carefully :)

**Comment by barrettcw** [comment]
This makes sense to me, but I would feel better if at least one of {Guy,
Tim} would agree...

-Clark

**Comment by guykatzz** [comment]
My reply is similar to the previous comments: the change makes sense to me, but my familiarity with this part of the code is limited.

File: src/prop/minisat/core/Solver.cc [hunk 1]
```diff
@@ -428,7 +428,10 @@ bool Solver::addClause_(vec<Lit>& ps, bool removable, ClauseId& id)
             }
           }
           return ok;
-        } else return ok;
+        } else {
+          PROOF(id = ClauseIdUndef;);
```
Comment (timothy-king) [inline src/prop/minisat/core/Solver.cc]: Can you update the bug w/ a more complete stack trace? I am not sure why setting the id for the calling context helps.

**Comment by 4tXJ7f** [comment]
Stack trace:
```
#0  CVC4::Minisat::Solver::addClause_ (this=0x61900000c880, ps=..., removable=false, id=@0x7fffffff4360: 4294967293) at ../../../../../../src/prop/minisat/core/Solver.cc:432
#1  0x0000000002cf26ff in CVC4::Minisat::SimpSolver::addClause_ (this=0x61900000c880, ps=..., removable=false, id=@0x7fffffff4360: 4294967293) at ../../../../../../src/prop/minisat/simp/SimpSolver.cc:179
#2  0x0000000002c020b1 in CVC4::Minisat::SimpSolver::addClause (this=0x61900000c880, ps=..., removable=false, id=@0x7fffffff4360: 4294967293) at ../../../../../../src/prop/minisat/../../prop/minisat/simp/SimpSolver.h:188
#3  0x0000000002bfecc9 in CVC4::prop::MinisatSatSolver::addClause (this=0x61500002e700, clause=std::vector of length 1, capacity 1 = {...}, removable=false) at ../../../../../../src/prop/minisat/minisat.cpp:155
#4  0x00000000021d1f43 in CVC4::prop::CnfStream::assertClause (this=0x6120000097c0, node=..., c=std::vector of length 1, capacity 1 = {...}) at ../../../../src/prop/cnf_stream.cpp:84
#5  0x00000000021d26fc in CVC4::prop::CnfStream::assertClause (this=0x6120000097c0, node=..., a=...) at ../../../../src/prop/cnf_stream.cpp:100
#6  0x00000000021e0675 in CVC4::prop::TseitinCnfStream::convertAndAssert (this=0x6120000097c0, node=..., negated=false) at ../../../../src/prop/cnf_stream.cpp:745
#7  0x00000000021e5462 in CVC4::prop::TseitinCnfStream::convertAndAssert (this=0x6120000097c0, node=..., removable=false, negated=false, proof_id=CVC4::RULE_GIVEN, from=...) at ../../../../src/prop/cnf_stream.cpp:697
#8  0x00000000022011cc in CVC4::prop::PropEngine::assertFormula (this=0x60b00000aac0, node=...) at ../../../../src/prop/prop_engine.cpp:129
#9  0x0000000000b8c1a0 in CVC4::smt::SmtEnginePrivate::processAssertions (this=0x61c00000c080) at ../../../../src/smt/smt_engine.cpp:4427
#10 0x0000000000b91876 in CVC4::SmtEngine::internalPush (this=0x619000007d80) at ../../../../src/smt/smt_engine.cpp:5478
#11 0x0000000000b8fdf4 in CVC4::SmtEngine::checkSatisfiability (this=0x619000007d80, ex=..., inUnsatCore=true, isQuery=false) at ../../../../src/smt/smt_engine.cpp:4523
#12 0x0000000000b8f068 in CVC4::SmtEngine::checkSat (this=0x619000007d80, ex=..., inUnsatCore=true) at ../../../../src/smt/smt_engine.cpp:4485
#13 0x0000000000acf28c in CVC4::CheckSatCommand::invoke (this=0x60800000b420, smtEngine=0x619000007d80) at ../../../../src/smt/command.cpp:297
#14 0x0000000000acbc7b in CVC4::Command::invoke (this=0x60800000b420, smtEngine=0x619000007d80, out=...) at ../../../../src/smt/command.cpp:109
#15 0x000000000050cdab in CVC4::main::smtEngineInvoke (smt=0x619000007d80, cmd=0x60800000b420, out=0x4f17260 <std::cout@@GLIBCXX_3.4>) at ../../../../../src/main/command_executor.cpp:179
#16 0x000000000050adb6 in CVC4::main::CommandExecutor::doCommandSingleton (this=0x611000009140, cmd=0x60800000b420) at ../../../../../src/main/command_executor.cpp:104
#17 0x000000000050a58f in CVC4::main::CommandExecutor::doCommand (this=0x611000009140, cmd=0x60800000b420) at ../../../../../src/main/command_executor.cpp:87
#18 0x0000000000515a54 in runCvc4 (argc=3, argv=0x7fffffffe5b8, opts=...) at ../../../../../src/main/driver_unified.cpp:535
#19 0x00000000005093a6 in main (argc=3, argv=0x7fffffffe5b8) at ../../../../../src/main/main.cpp:55
```

Minimal example:
```
(set-option :produce-unsat-cores true)
(set-logic UF)
(push 1)
(declare-fun _substvar_4_ () Bool)
(assert _substvar_4_)
(assert _substvar_4_)
(check-sat)
```

**Comment by timothy-king** [comment]
Let's add that minimized example into test/regress/regress0/push-pop/.

**Comment by 4tXJ7f** [comment]
Sorry, I wanted to rebase but I could not force push anymore (did we disable that?), so I had to merge, which is why the pull request looks messy. I added the test case.

**Comment by 4tXJ7f** [comment]
Opened a new PR (#180).

**Pull request closed without merging**